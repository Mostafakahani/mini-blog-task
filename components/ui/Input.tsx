import React from "react";

interface InputProps {
  id: string;
  label: string;
  type?: string;
  value: string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  required?: boolean;
  className?: string;
}

const Input = ({
  id,
  label,
  type = "text",
  value,
  onChange,
  required = false,
  className = "",
}: InputProps) => {
  return (
    <div className="mb-4">
      <label
        className="block text-gray-700 text-sm font-bold mb-2"
        htmlFor={id}
      >
        {label}
      </label>
      <input
        id={id}
        type={type}
        className={`w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 ${className}`}
        value={value}
        onChange={onChange}
        required={required}
      />
    </div>
  );
};

export default Input;
