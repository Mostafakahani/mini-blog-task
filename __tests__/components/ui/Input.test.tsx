import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Input from "@/components/ui/Input";

describe("Input", () => {
  it("renders the label and input element", () => {
    render(
      <Input id="username" label="Username" value="" onChange={() => {}} />
    );
    expect(screen.getByLabelText("Username")).toBeInTheDocument();
  });

  it("passes the type prop to the input element", () => {
    render(
      <Input
        id="email"
        label="Email"
        type="email"
        value=""
        onChange={() => {}}
      />
    );
    expect(screen.getByLabelText("Email")).toHaveAttribute("type", "email");
  });

  it("calls onChange when input value changes", () => {
    const handleChange = jest.fn();
    render(
      <Input id="password" label="Password" value="" onChange={handleChange} />
    );
    const inputElement = screen.getByLabelText("Password");
    fireEvent.change(inputElement, { target: { value: "new value" } }); // simulate input change
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it("applies the required attribute when required prop is true", () => {
    render(
      <Input id="name" label="Name" value="" onChange={() => {}} required />
    );
    expect(screen.getByLabelText("Name")).toBeRequired();
  });

  it("includes additional className prop in the input element", () => {
    render(
      <Input
        id="custom"
        label="Custom Input"
        value=""
        onChange={() => {}}
        className="custom-class"
      />
    );
    expect(screen.getByLabelText("Custom Input")).toHaveClass("custom-class");
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Input
        id="snapshot"
        label="Snapshot Input"
        value="test"
        onChange={() => {}}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
