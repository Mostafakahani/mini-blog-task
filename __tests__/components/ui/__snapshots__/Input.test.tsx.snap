// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches snapshot 1`] = `
<div
  class="mb-4"
>
  <label
    class="block text-gray-700 text-sm font-bold mb-2"
    for="snapshot"
  >
    Snapshot Input
  </label>
  <input
    class="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 "
    id="snapshot"
    type="text"
    value="test"
  />
</div>
`;
